x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/plan/QueryPlan.java

```diff
@@ -52,22 +52,36 @@ public AttributeSet inputSet() {
         return lazyInputSet;
     }
 
-    public PlanType transformExpressionsOnly(Function<? super Expression, ? extends Expression> rule) {
-        return transformPropertiesOnly(e -> doTransformExpression(e, exp -> exp.transformDown(rule)), Object.class);
+    //
+    // pass Object.class as a type token to pick Collections of expressions not just expressions
+    //
+
+    public PlanType transformExpressionsOnly(Function<Expression, ? extends Expression> rule) {
+        return transformPropertiesOnly(rule, Expression.class);
     }
 
-    public PlanType transformExpressionsDown(Function<? super Expression, ? extends Expression> rule) {
-        return transformPropertiesDown(e -> doTransformExpression(e, exp -> exp.transformDown(rule)), Object.class);
+    public <E extends Expression> PlanType transformExpressionsOnly(Function<E, ? extends Expression> rule, Class<E> typeToken) {
```

Comment:
It would however the current pattern is to pass the token class last.

Reply:
I'd also agree with this, since many (most?) rules are lambdas. Maybe it might be worth at some point adding redundant methods with the arguments swapped and slowly migrate to using those -- just an idea.